### PR TITLE
Add meta_init, enable it as default init process

### PR DESCRIPTION
**Pull request by lessw2020**
This PR enables meta_init functionality to avoid OOM'ing on cpu for larger models. 
The core functionality is in meta_init.py, and a few changes in parallelization and train.py. 
Key items:
1 - this is largely the same as the earlier PR I had for meta_init, but I did a new one b/c faster than reworking it with all the interim changes. 
2 - to address feedback in previous PR:
a - why do we need meta_init.py, can't we just do:
~~~
with torch.device("meta"):
    model = Model.from_args(...)
~~~
Unfortunately this does not work b/c the rope embeddings are treated differently (buffer) and thus the simple lambda call from param_init_fn in FSDP  (lambda module: module.to_device('cuda') ) will not invoke or move the rope embeddings and the model will fail on first forward.
This issue relates to the nn.embeddings not being moved, and that the device is referenced in the forward pass for the current rope class.  Have opened https://github.com/pytorch/torchtrain/issues/110 to track this and investigate while not holding up meta init that is working from landing.

b - per earlier feedback - meta init is now 'not optional' but simply the default.  This should ensure all models leverage it and ensure we aren't missing things for future meta_init aspects. 

3 - misc change - I switched the model_params to just do the normal all params count instead of 'unique params' b/c it does not mesh with what people perceive model size as.  

Testing:
tested both debugmodel and 26B model with and without meta init to confirm same loss curves.  
Note for future reference - if you get a bad init (meta init failure) you will simply not train (loss is same every iter). 
If you fail to call reset params after FSDP, then you will train (b/c we default to torch.randn_like) but your starting loss will be 5x+ higher (telling you that you have not properly init'ed the model).

### CODE REVIEW

File: torchtrain/models/llama/__init__.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "1B": ModelArgs(dim=1024, n_layers=16, n_heads=8),
     "7B": ModelArgs(dim=4096, n_layers=32, n_heads=32),
     "13B": ModelArgs(dim=5120, n_layers=40, n_heads=40),
-    "40B": ModelArgs(dim=5120, n_layers=80, n_heads=40),
+    "26B": ModelArgs(dim=5120, n_layers=80, n_heads=40),
```
Comment (wanchaol) [inline torchtrain/models/llama/__init__.py:14]: why we change this to 26B, is it because the previous calculation of 40B is wrong?

File: train.py [hunk 2]
```diff
@@ -128,6 +131,10 @@ def main(job_config: JobConfig):
         model, world_mesh, parallel_dims, job_config
     )
 
+    # we have now moved from meta to device,
+    # reset parameters for proper initialization
+    model.reset_parameters()
```
Comment (wanchaol) [inline train.py]: maybe we should call this inside `parallelize_llama` instead?

File: torchtrain/parallelisms/parallelize_llama.py [hunk 3]
```diff
@@ -164,23 +166,21 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
                 model.layers[layer_id] = wrap(transformer_block)
 
             # wrap the rest layers with FSDP
-            model = wrap(model.cuda())
+            model = wrap(model)
 
         rank0_log("Applied FSDP to the model...")
 
-    # redundant if FSDP is enabled, but ensure the model is on device regardless of which parallelisms were used
-    model.cuda()
```
Comment (tianyu-l) [inline torchtrain/parallelisms/parallelize_llama.py:228]: I wonder if we should keep this for models with `dp_enabled=False`?

File: torchtrain/metrics.py [hunk 4]
```diff
@@ -192,8 +192,8 @@ def get_num_params(model: nn.Module, only_trainable: bool = False) -> int:
     param_list = list(model.parameters())
     if only_trainable:
         param_list = [p for p in param_list if p.requires_grad]
-    unique_params = {p.data_ptr(): p for p in param_list}.values()
-    return sum(p.numel() for p in unique_params)
+    # unique_params = {p.data_ptr(): p for p in param_list}.values()
+    return sum(p.numel() for p in param_list)
```
Comment (tianyu-l) [inline torchtrain/metrics.py:197]: FYI by running on the 1B flavor llama model, I got the same output both before and after this change: "Model llama 1B size: 271,090,688 total parameters". Is this expected?

Comment (tianyu-l) [reply torchtrain/metrics.py:197]: Btw, with 7B flavor, I got accurate "Model llama 7B size: 6,738,415,616 total parameters".

File: torchtrain/models/llama/__init__.py (hunk 1 above)
Comment (lessw2020) [reply torchtrain/models/llama/__init__.py:14]: yes - after I added the num params check and updated it to reflect all params (not unique) it showed the 40B was really 26B so I just updated it to be correct.

File: torchtrain/metrics.py (hunk 4 above)
Comment (lessw2020) [reply torchtrain/metrics.py:197]: yes, it's a labelling mistake rather than a param count issue. 
I've updated the specs so now we have:
271M
and a correct 1B option:
"Model llama 1B size: 1,055,991,808 total parameters"

File: torchtrain/parallelisms/parallelize_llama.py (hunk 3 above)
Comment (lessw2020) [reply torchtrain/parallelisms/parallelize_llama.py:228]: re: error out if meta init fails - 
In this error case, you'll have some zero'ed out weight params which is why it won't train (no gradients propagating due to the zeros).
I don't think there's a lightweight way to catch that without traversing the entire model looking for zero'ed weights and not sure that overhead of running that for every init as it would/should be rare vs users simply being aware of what happens in this case would be worth it, esp for large models. 
If we start seeing people reporting their models won't train then I'd say we can revisit and add a verify meta init option.

Comment (lessw2020) [reply torchtrain/parallelisms/parallelize_llama.py:228]: re: keep model.cuda() for dp_enabled = False
good catch.  I've added the direct model.cuda() move back for the non_dp case.

File: train.py (hunk 2 above)
Comment (lessw2020) [reply train.py]: I put the reset_params call in the training loop directly, b/c regardless of parallelization plan, we always want to call this.  
i.e. if we later have parallelize_mixtral, parallelize_gpt, etc. then if it's in parallelize_llama everybody has to repeat this reset_params call, vs only being once after any parallellize plan was called.

Since we don't have any other parallelize plans atm, then I went ahead and moved it inside parallelize llama since for now it's cleaner...but wanted to clarify the logic behind the original placement.

File: torchtrain/models/llama/model.py [hunk 5]
```diff
@@ -466,7 +466,14 @@ def __init__(self, model_args: ModelArgs):
...
 
         # init model weights
-        self.reset_parameters()
+
+        # we are doing meta_init, which will call reset_parameters() after
+        # the model is moved to actual device.
+        # If you modify and are not using meta_init, you will need to call
+        # reset_parameters() manually as below:
+
+        # self.reset_parameters()
```
Comment (tianyu-l) [inline torchtrain/models/llama/model.py:477]: Do you think it makes sense to remove the `reset_parameters()` within the `__init__` of `RMSNorm` as this will make code look consistent.

Comment (lessw2020) [reply torchtrain/models/llama/model.py:477]: yes, good catch! updated.

File: torchtrain/parallelisms/parallelize_llama.py [hunk 6]
```diff
@@ -220,10 +221,13 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
 
         rank0_log("Applied FSDP to the model...")
+    else:
+        model.cuda()
 
-    # redundant if FSDP is enabled, but ensure the model is on device regardless of which parallelisms were used
-    model.cuda()
+    # we have now moved from meta to device,
+    # reset parameters for proper initialization
+    model.reset_parameters()
```
Comment (awgu) [inline torchtrain/parallelisms/parallelize_llama.py:232]: I apologize that FSDP meta-device init is confusing, but I think this might not be fully correct.

1. The contract from PyTorch core is supposed to be that `module.reset_parameters()` only resets/initializes the parameters immediately owned by `module`, not those of its children/submodules. Here, `model: Transformer` is initializing all of its submodules' parameters. This contract is the only way for `reset_parameters()` to always work compositionally, otherwise (like on our case) we _must_ assume `Transformer` to always be the root module.
2. When we call `model.reset_parameters()`, the parameters have already been flattened and sharded by FSDP. This means that any initialization method that depends on the tensor shape would be incorrect. That is why we would normally want users to do the correct initialization in the `param_init_fn`.

For this Llama case, it looks like perhaps the `reset_parameters()` have been written to not depend on the tensor shape directly?

Comment (lessw2020) [reply torchtrain/parallelisms/parallelize_llama.py:232]: correct that the init does not depend on tensor shape directly.
I see your point though about only resetting it's own params and not children, but let's meet to discuss as I also have questions on how meta_init should work for FSDP2 and we can get an updated impl.
